contrib/mesos/pkg/scheduler/podtask/pod_task.go

```diff
@@ -0,0 +1,365 @@
+	StateRunning
+	StateFinished
+	StateUnknown
+)
+
+type FlagType string
+
+const (
+	Launched = FlagType("launched")
+	Bound    = FlagType("bound")
+	Deleted  = FlagType("deleted")
+)
+
+// A struct that describes a pod task.
+type T struct {
+	ID          string
+	Pod         api.Pod
+	Spec        Spec
+	Offer       offers.Perishable // thread-safe
+	State       StateType
+	Flags       map[FlagType]struct{}
+	CreateTime  time.Time
+	UpdatedTime time.Time // time of the most recent StatusUpdate we've seen from the mesos master
+
+	podStatus  api.PodStatus
+	executor   *mesos.ExecutorInfo // readonly
+	podKey     string
+	launchTime time.Time
+	bindTime   time.Time
+	mapper     HostPortMappingType
+}
+
+type Spec struct {
+	SlaveID string
+	CPU     float64
+	Memory  float64
+	PortMap []HostPortMapping
+	Ports   []uint64
+	Data    []byte
+}
+
+// mostly-clone this pod task. the clone will actually share the some fields:
+//   - executor    // OK because it's read only
+//   - Offer       // OK because it's guarantees safe concurrent access
+func (t *T) Clone() *T {
+	if t == nil {
+		return nil
+	}
+
+	// shallow-copy
+	clone := *t
+
+	// deep copy
+	(&t.Spec).copyTo(&clone.Spec)
+	clone.Flags = map[FlagType]struct{}{}
+	for k := range t.Flags {
+		clone.Flags[k] = struct{}{}
+	}
+	return &clone
+}
+
+func (old *Spec) copyTo(new *Spec) {
+	if len(old.PortMap) > 0 {
+		new.PortMap = append(([]HostPortMapping)(nil), old.PortMap...)
+	}
+	if len(old.Ports) > 0 {
+		new.Ports = append(([]uint64)(nil), old.Ports...)
+	}
+	if len(old.Data) > 0 {
+		new.Data = append(([]byte)(nil), old.Data...)
+	}
+}
+
+func (t *T) HasAcceptedOffer() bool {
+	return t.Spec.SlaveID != ""
+}
+
+func (t *T) GetOfferId() string {
+	if t.Offer == nil {
+		return ""
+	}
+	return t.Offer.Details().Id.GetValue()
+}
+
+func generateTaskName(pod *api.Pod) string {
+	ns := pod.Namespace
+	if ns == "" {
+		ns = api.NamespaceDefault
+	}
+	return fmt.Sprintf("%s.%s.pods", pod.Name, ns)
+}
+
+func (t *T) BuildTaskInfo() *mesos.TaskInfo {
+	info := &mesos.TaskInfo{
+		Name:     proto.String(generateTaskName(&t.Pod)),
+		TaskId:   mutil.NewTaskID(t.ID),
+		SlaveId:  mutil.NewSlaveID(t.Spec.SlaveID),
+		Executor: t.executor,
+		Data:     t.Spec.Data,
+		Resources: []*mesos.Resource{
+			mutil.NewScalarResource("cpus", t.Spec.CPU),
+			mutil.NewScalarResource("mem", t.Spec.Memory),
+		},
+	}
+	if portsResource := rangeResource("ports", t.Spec.Ports); portsResource != nil {
+		info.Resources = append(info.Resources, portsResource)
+	}
+	return info
+}
+
+// Fill the Spec in the T, should be called during k8s scheduling,
+// before binding.
+func (t *T) FillFromDetails(details *mesos.Offer) error {
+	if details == nil {
+		//programming error
+		panic("offer details are nil")
+	}
+
+	log.V(3).Infof("Recording offer(s) %v against pod %v", details.Id, t.Pod.Name)
+
+	t.Spec = Spec{
+		SlaveID: details.GetSlaveId().GetValue(),
+		CPU:     containerCpus,
+		Memory:  containerMem,
```

Comment:
Actually, never mind this comment. I see that you're using containerCpus and containerMem as defaults. What I actually don't understand is the caller, not this function. (See my comment where this function is called.)